Coding Challenge Refresh (#1)
noyo-technologies/backend-coding-challenge: README.md

# Table of Contents
- [Table of Contents](#table-of-contents)
  - [TLDR;](#tldr)
- [Noyo Coding Challenge! (Introduction)](#noyo-coding-challenge-introduction)
  - [Problem 1. Implement Feature (~ 2 Hours)](#problem-1-implement-feature--2-hours)
    - [Business Rules](#business-rules)
- [What we are evaluating](#what-we-are-evaluating)
- [What we are not evaluating](#what-we-are-not-evaluating)
- [Cheat Sheet](#cheat-sheet)
- [Submission Options](#submission-options)
  - [Support](#support)
  - [Live Extensions](#live-extensions)
  - [Instructions For Running](#instructions-for-running)
    - [Building the App](#building-the-app)
    - [Running the App](#running-the-app)
    - [Running the Tests](#running-the-tests)
    - [Seeding the Database](#seeding-the-database)
    - [Connecting to the database](#connecting-to-the-database)
    - [Formatting your Code](#formatting-your-code)
    - [Debugging](#debugging)
    - [Starting Over](#starting-over)
- [Thanks!](#thanks)

## TLDR;

> 1. You'll need docker, install it.

Reviewer: ```suggestion
1. We use Docker for our local development here at Noyo, you can find the instructions for installing it on your local system [here](https://docs.docker.com/get-docker/).
```

> 2. Run the following... and 12 test will pass, with no warnings or errors.

Reviewer 2: ```suggestion
2. Run the following... and 12 tests will pass, with no warnings or errors.
```

      ```sh

>       docker compose build # build some stuff
>       docker compose up -d && sleep 5 # start the server and wait
>       docker-compose exec service python seed.py # seed the db 
>       docker compose exec service pytest . # run some tests

Reviewer: ```suggestion
      docker compose build # build the local containers
      docker compose up -d && sleep 5 # start up the Docker containers and wait
      docker-compose exec service python seed.py # Seed the database
      docker compose exec service pytest . # Run the test suite
```

      ```

> 3. Do [Take Home Problem 1](#problem-1-implement-feature--2-hours)

Reviewer: ```suggestion
3. Complete [Take Home Problem 1](#problem-1-implement-feature--1-2-hours)
```

4. If you need to start over
   ```sh

>    docker-compose down --remove-orphans --volumes # nuke all the things

Reviewer: ```suggestion
   docker-compose down --remove-orphans --volumes # stop all containers from running and remove volumes
```

   ```

> 5. Look at the [Live Extensions](#live-extensions), if you want, but dont do them.

Reviewer: ```suggestion
5. Look at the [Live Extensions](#live-extensions), if you want, but dont do them.
5. Feel free to familiarize yourself with the [Live Extensions](#live-extensions), but please do not implement them.
```

# Noyo Coding Challenge! (Introduction)

> Within this repository is a python application that implements a fictitious product specification. Noyo is partnering with innovative insurance companies offering a new type of coverage called "traveling salesperson insurance.". This offering gives a company's employees unique benefits when they are on the road, it considers various aspects of their travel and has some very complex plan eligibility rules. When a salesman is working out of a home office, there could be gaps in their coverage timeline. 

Reviewer: ```suggestion
Within this repository is a Python application that implements a fictitious product specification. Noyo is partnering with an innovative insurance company that is offering a new type of coverage called Traveling Salesperson Insurance. This offering gives a company's employees unique benefits when they are on the road. It considers various aspects of their travel and has some complex plan eligibility rules. When a salesperson is working out of a home office, there could be gaps in their coverage timeline. 
```

This code was deployed to production as is but is incomplete. The original engineer was on a tight timeline and did their best; there may be bugs in the code in addition to insufficient or incorrect tests.  

**Note**  
There are **no intentionally** included bugs that you have to seek out or find for this challenge. However, the incomplete tests are intentional. If you see something that you think is a bug, let's talk about it during your onsite!

> **Part 1 (Take home - Completed before the interview) ~ 2 Hours**  

Reviewer: ```suggestion
**Part 1 (Take home - Completed before the interview) 1-2 Hours**  
```

Development has halted and the product is incomplete!  
1. Complete development as outlined in the challenge. 

**Part 2 (Onsite - Pair programming session) ~ 75 Minutes**  
An engineer has joined your team, show them the ropes!  
1. Discuss the take-home portion in Part 1
2. Live code implementation of additional features.

## Problem 1. Implement Feature (~ 2 Hours)

For this exercise, you'll add functionality to the `PUT /api/persons/<person_id>/segment` endpoint, which creates a `Segment` for the `Person` with the specified ID.

To begin, you will need to [start the server](#running-the-app) and create some `Person` records either by making calls to the server or running the [seed](#seeding-the-database) helper we've provided.

> After you've created at least one `Person`, plug their `id` into the following `curl` command (or construct an API call using the tool of your choice).

Reviewer: ```suggestion
After you've created at least one `Person`, plug their `id` into the following `curl` command (or construct an API call using the tool of your choice like [Postman](https://www.postman.com/)).
```

```bash
curl -X PUT \
  http://localhost:3000/api/persons/<person_id>/segment \
  -H "Content-Type: application/json" \
  -d '{
    "start_date": "2021-01-01",
    "city": "San Francisco",
    "state": "CA",
    "zip_code": "94613"
}'
```

> _The previous API call will produce a 501 not implemented._ However its should create a single `Segment` record for the `Person` with a `start_date` of `2021-01-01` and an `end_date` of `null`. In our data model, this means that the person's location was `San Francisco, CA 94613` starting on `2021-01-01` and will remain that indefinitely. The following diagram is a visual representation that:

Reviewer: ```suggestion
_The previous API call will produce a 501 not implemented._ However, what it should do is to create a single `Segment` record for the `Person` with a `start_date` of `2021-01-01` and an `end_date` of `null`. In our data model, this means that the person's location was `San Francisco, CA 94613` starting on `2021-01-01` and will remain that indefinitely. The following diagram is a visual representation that:
```

![Figure 1](/docs/figure_1.png)

> Making another API call that updates the person's location to `San Francisco, CA 94911` starting on `2021-04-15` and ending one `2021-08-15`. Should add this new segment to the person's existing segment list.

Reviewer 2: ```suggestion
Making another API call that updates the person's location to `San Francisco, CA 94911` starting on `2021-04-15` and ending one `2021-08-15`, should add this new segment to the person's existing segment list.
```

  http://localhost:3000/api/persons/8abbc8e0-7039-4988-8a20-e6c1e1b3bb2f/segment \
  -H "Content-Type: application/json" \
  -d '{
    "start_date": "2021-04-15",
    "end_date": "2021-08-15",
    "city": "San Francisco",
    "state": "CA",
    "zip_code": "94911"
  }'
```
![Figure 2](/docs/figure_2.png)

Making a third API call that updates the person's location to `Houston, TX 12345` starting on `2021-09-15`. Should add this new segment to the person's existing segment list.

```bash
curl -X PUT \
  http://localhost:3000/api/persons/8abbc8e0-7039-4988-8a20-e6c1e1b3bb2f/segment \
  -H "Content-Type: application/json" \
  -d '{
    "start_date": "2021-09-15",    
    "city": "Houston",
    "state": "TX",
    "zip_code": "12345"
  }'
```
![Figure 3](/docs/figure_3.png)

A final API call that updates the person's location to `Durham, NC 27517` starting on `2021-07-15`. Should fail.

```bash
curl -X PUT \
  http://localhost:3000/api/persons/8abbc8e0-7039-4988-8a20-e6c1e1b3bb2f/segment \
  -H "Content-Type: application/json" \
  -d '{
    "start_date": "2021-07-15",    
    "city": "Durham",
    "state": "NC",
    "zip_code": "27517"
  }'
```
![Figure 4](/docs/figure_4.png)

> Your challenge is to update the in segment.py to `create_segment` handling the creation of location segments. 

Reviewer: ```suggestion
Your challenge is to update `create_segment` method in the `segment.py` file to handle the creation of location segments.
```

### Business Rules

> 1. You can only add segments to the end.
>     1. That is, adding a segment whose start date is before any current start date should result in a 422.
> 2. You should treat a segment's end date of None as indefinitely in the future.
> 3. A new segment can have a non-None end date.
> 4. Update the end_date of an existing segment to the start_date of the new segment.
>     1. If the existing segment end_date is before the new segment's start_date, no update to the current segment is required.
> 5. The endpoint should implement HTTP PUT semantics when inserting exact duplicates.
> 6. If a person does not exist, return 404

Reviewer: ```suggestion
1. You can only add segments to the end. That is, adding a segment whose start date is before any current start date should result in a `422`.
2. You should treat a segment's `end_date` of `None` as indefinitely into the future.
3. A new segment can have an `end_date` that is not `None`
4. If the `start_date` of the new segment is less than the `end_date` of the existing segment (or the `end_date` is `None`) update the the existing segment's `end_date` to match the `start_date` of the new segment.
5. The endpoint should implement [HTTP PUT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT) semantics when inserting exact duplicates.
6. If a person does not exist, return `404`
```

# What we are evaluating

Whenever you start a new position, you encounter code other people have written. You'll have to understand the architecture, write code to match it, and improve it for the next engineer. We are looking for:

- How do you approach writing a new feature? 
- Does your code do what it's supposed to do without obvious errors or bugs?
- Can you debug a problem in some unfamiliar code? How did you go about debugging?
- Can you evaluate, review, and improve upon code that is either poor or perhaps difficult to understand and maintain.

*💡 There are no right or wrong answers to the concepts above, but we're interested in your thinking, arguments & results.*

# What we are not evaluating

> - Obtuse or complex python language features or skills.

Reviewer: ```suggestion
- Obtuse or complex Python language features or skills.
```

> We are a python shop and expect you to have a basic understanding of python language features (or can learn them quickly), particularly those found in many other modern languages. 

Reviewer: ```suggestion
We primarily develop using Python here at Noyo, we expect you to have a basic understanding of the Python language features (or can learn them quickly), particularly those found in many other modern languages. 
```

Hint: Brush up on classes, functions, loops, if-else control flow, iterators, dictionaries, lists, tuples, datetime.date

> *This challenge should not require in-depth python knowledge or a deep understanding of any third-party Library. To that end, here are is a quick [Cheat sheet](/docs/cheat_sheet.md) designed to get you over hurdles that are not particularly part of this evaluation*

Reviewer: ```suggestion
*This challenge should not require in-depth Python knowledge or a deep understanding of any third-party Library. To that end, here are is a quick [Cheat sheet](/docs/cheat_sheet.md) designed to get you over hurdles that are not particularly part of this evaluation*
```

# Cheat Sheet
[Cheat sheet](/docs/cheat_sheet.md)

# Submission Options

You have two options for completing the exercise; choose the one that you are most comfortable with:

1. Zip up your code and email it to coding-challenge@noyo.com

2. Commit your code to a Github repository and share the link to the code in an email to coding-challenge@noyo.com

- **Note: Please do not fork this repository or open a pull request with your submission against this repository.**

## Support

If you have any issues or questions about the instructions below, please email **coding-challenge@noyo.com**, and we will get back to you as soon as possible.

## Live Extensions

> We will ask you to implement one or more extensions to the code during your live coding interview. You are welcome to look at them, and encouraged to do so. However, you need not spend additional time implementing them beforehand.

Reviewer: ```suggestion
We will ask you to implement one or more extensions to the code during your live coding interview. You are welcome to look at them, and encouraged to do so. However, please do not spend additional time implementing them beforehand.
```

```bash

docker compose build

```

### Running the App

You can start the server and database using the following command:

```bash

docker compose up -d

```

This will start up the [Python Flask](https://flask.palletsprojects.com/en/1.1.x/) server on `localhost:3000` and a PostgreSQL instance on `localhost:5432`.

Flask starts up in debug mode, meaning whenever you save a file, it will restart the server with your changes.

### Running the Tests

We have provided a test suite that uses [pytest](https://docs.pytest.org/en/stable/). You can run the test suite using the following:

```bash

docker compose exec service pytest .

```

- *Note: The tests use the same database as the local server. When you run tests, the database is cleared of all data.**

### Seeding the Database

To make it easier to work with the API, we've provided a simple Python script that you can run to create five random `person` entries in the database. You can run the seed using the following:

```bash

docker-compose exec service python seed.py

```

- *Note: You can only run the seed after you have started the server using `docker-compose up`*

> ### Connecting to the database

Reviewer: ```suggestion
### Connecting to the Database
```
